2-collectors/scala-stream-collector/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -255,19 +286,76 @@ class KinesisSink private (
                   s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]"
                 )
             )
-            log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
-            val failures = failurePairs.map(_._1)
-            scheduleBatch(failures, nextBackoff)
+            maybeSqs match {
+              case Some(sqs) =>
+                log.info(s"Sending all failed records to SQS buffer queue: ${sqs.sqsBufferName}")
+                putToSqs(sqs, failurePairs.map(_._1))
+              case None =>
+                log.error(
+                  s"Dropping ${failurePairs.size} failed events (consider setting a SQS Buffer in config.hocon)"
+                )
+            }
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          maybeSqs match {
+            case Some(sqs) =>
+              log.info(
+                s"Sending all (${batch.size}) events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+              )
+              putToSqs(sqs, batch)
+            case None =>
+              log.error(
+                s"Dropping all (${batch.size}) events from a batch (consider setting a SQS Buffer in config.hocon)"
+              )
+          }
         }
       }
     }
 
+  private def putToSqs(sqs: SqsClientAndName, batch: List[(ByteBuffer, String)]): Future[Unit] =
+    Future {
+      log.info(s"Writing ${batch.size} messages to SQS queue: ${sqs.sqsBufferName}")
+      val encoded = batch.map {
+        case (msg, key) =>
+          // The key is either IP or random UUID, so neither should contin `|`.
+          // Sqs doesn't support keys, so `|` is used as a separator. The same needs to be set up in
```

Comment:
`|` is not hard coded and can be anything

Reply:
It's a default separator for now, so unless someone overwrite it it defaults to `|`

Reply:
I also have a better idea how to implement it, so I wold remove the whole separator away.

Reply:
Great!